corehq/apps/builds/utils.py

```diff
@@ -40,3 +44,61 @@ def extract_build_info_from_filename(content_disposition):
     else:
         raise ValueError('Could not find filename like {!r} in {!r}'.format(
             pattern, content_disposition))
+
+
+@memoized
+def get_latest_version_at_time(config, target_time):
+    """
+    Get the latest CommCare version that was available at a given time.
+    Excludes superuser-only versions.
+    Menu items are already in chronological order (newest last).
+    If no target time is provided, return the latest version available now.
+
+    Args:
+        config: CommCareBuildConfig instance
+        target_time: datetime or string in ISO format, or None for latest version
+    """
+    if not target_time:
+        return config.get_default().version
+
+    if isinstance(target_time, str):
+        target_time = datetime.strptime(target_time, ISO_DATETIME_FORMAT)
+
+    # Iterate through menu items in reverse (newest to oldest)
+    for item in reversed(config.menu):
+        if item.superuser_only:
+            continue
+        try:
+            build_time = get_build_time(item.build.version)
+            if build_time and build_time <= target_time:
+                return item.build.version
+        except KeyError:
+            continue
+
+    return None
+
+
+@memoized
+def get_build_time(version):
+    build = CommCareBuild.get_build(version, latest=True)
+    if build and build.time:
+        return build.time
+    return None
```

Comment:
Thank you! Sorry this is being extracted into its own function and memoized in a rush. Fixed in bb87820

Reply:
Nit: Your version is an improvement, and it hides the exception complexity, but it still does more than is needed in the try/except block. `CommCareBuild` is the statement that can throw, so its slightly better and more explicit to write:
```python
try:
  build = CommCareBuild.get_build(version, latest=True)
except KeyError:
  return None
return build.time if build and build.time else None
```

because it scopes the exception block to _just_ the code that can raise the exception. I also prefer to return an explicit value, rather than letting it implicitly return nothing.

Reply:
84f1080d10515d9361af2478937078c172128a93